### PR TITLE
[mtouch] Unify user resources removal as a link step

### PR DESCRIPTION
**Pull request by spouliot**
Right now the logic exists in a few places, both in and outside the
linker. We recently began to use part of the linker pipeline in normal /
all builds so it's easier to share (and unify) the code now.

The real gain is to avoid copying assemblies, in particular large ones,
more than strictly needed while building.

E.g. a build including a very large 1.3GB assembly, with several
native libraries embedded, save a lot of time avoiding the rewrites

mtouch (before)
        Total time: 64202 ms

mtouch (after)
        Total time: 34840 ms

### CODE REVIEW

File: tools/linker/RemoveUserResourcesSubStep.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			
+			HashSet<string> libraries = null;
+			if (assembly.HasCustomAttributes) {
+				foreach (var ca in assembly.CustomAttributes) {
+					if (!ca.AttributeType.Is ("ObjCRuntime", "LinkWithAttribute"))
+						continue;
+					var lwa = Xamarin.Bundler.Assembly.GetLinkWithAttribute (ca);
+					if (libraries == null)
+						libraries = new HashSet<string> (StringComparer.OrdinalIgnoreCase);
+					libraries.Add (lwa.LibraryName);
```
Comment (rolfbjarne) [inline tools/linker/RemoveUserResourcesSubStep.cs]: `lwa.LibraryName` might be null.

File: tools/linker/RemoveUserResourcesSubStep.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+				}
+			}
+
+			if (libraries == null)
+				return;
```
Comment (rolfbjarne) [inline tools/linker/RemoveUserResourcesSubStep.cs]: Assemblies can have MonoTouch resources (`__monotouch_content_` / `__monotouch_page_ `) without having `LinkWith` attributes, so I don't think this check is correct.

Comment (spouliot) [inline tools/linker/RemoveUserResourcesSubStep.cs]: Right, that should have been removed when I merged the existing step inside the new one.

File: tools/linker/RemoveUserResourcesSubStep.cs (hunk 1 above)
Comment (spouliot) [inline tools/linker/RemoveUserResourcesSubStep.cs]: good catch

**Comment by spouliot** [comment]
@chamons the removal of `tools/linker/MonoTouch.Tuner/RemoveMonoTouchResources.cs` affects `mmp` - but it seems this step _might_ not be working as it uses `__monotouch_*` constants and the msbuild tasks use a `Prefix` which I assume is not `monotouch` for XM.

**Comment by chamons** [comment]
LGTM assuming tests pass 👍

**Comment by rolfbjarne** [comment]
build

**Comment by rolfbjarne** [comment]
Many test failures, although all look unrelated:

1) MSBuild tests - iOS: ibtool hung (bug [#46097](https://bugzilla.xamarin.com/show_bug.cgi?id=46097))

2) monotouch-test/watchOS: failed to launch (bug [#45909](https://bugzilla.xamarin.com/show_bug.cgi?id=45909))

3) introspection/Unified and introspection/UnifiedXM crashed (bug [#46115](https://bugzilla.xamarin.com/show_bug.cgi?id=46115))

**Comment by spouliot** [comment]
build

**Pull request merged**